File: apps/mobile/src/inlineModules/AtomicFileModule.swift
import Darwin
internal import ExpoModulesCore
import Foundation

/// ADR V7-0005 durability primitive: fully-synchronized temp writes and atomic rename.
class AtomicFileModule: Module {
  func definition() -> ModuleDefinition {
    Name("AtomicFileModule")

    AsyncFunction("writeAndSync") { (url: URL, bytes: Data) in
      guard url.isFileURL else { throw AtomicFileError.nonFileURL }
      try FileManager.default.createDirectory(
        at: url.deletingLastPathComponent(),
        withIntermediateDirectories: true
      )
      if !FileManager.default.fileExists(atPath: url.path) {
        FileManager.default.createFile(atPath: url.path, contents: nil)
      }
      let handle = try FileHandle(forWritingTo: url)
      defer { try? handle.close() }
      try handle.truncate(atOffset: 0)
      try handle.write(contentsOf: bytes)
      // fsync/FileHandle.synchronize can return before the storage hardware has
      // committed its cache on Apple platforms. F_FULLFSYNC is the documented
      // power-loss durability primitive required by the V7-5 X3 obligation.
      if Darwin.fcntl(handle.fileDescriptor, F_FULLFSYNC) != 0 {
        throw POSIXError(POSIXErrorCode(rawValue: errno) ?? .EIO)
      }
    }

    AsyncFunction("replaceAndSync") { (source: URL, destination: URL) in
      guard source.isFileURL, destination.isFileURL else {
        throw AtomicFileError.nonFileURL
      }
      let result = source.path.withCString { sourcePath in
        destination.path.withCString { destinationPath in
          Darwin.rename(sourcePath, destinationPath)
        }
      }
      if result != 0 {
        throw POSIXError(POSIXErrorCode(rawValue: errno) ?? .EIO)
      }
      let directoryPath = destination.deletingLastPathComponent().path
      let descriptor = Darwin.open(directoryPath, O_RDONLY)
      if descriptor < 0 {
        throw POSIXError(POSIXErrorCode(rawValue: errno) ?? .EIO)
      }
      defer { Darwin.close(descriptor) }
      if Darwin.fsync(descriptor) != 0 {
        throw POSIXError(POSIXErrorCode(rawValue: errno) ?? .EIO)
      }
    }
  }
}

private enum AtomicFileError: Error {
  case nonFileURL
}

File: apps/mobile/src/inlineModules/AtomicFileModule.kt
package app

import android.net.Uri
import android.system.Os
import android.system.OsConstants
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream

/** Durability primitive for fsynced temporary writes and atomic rename. */
class AtomicFileModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("AtomicFileModule")

    AsyncFunction("writeAndSync") { uriString: String, bytes: ByteArray ->
      val file = fileFromUri(uriString)
      file.parentFile?.mkdirs()
      FileOutputStream(file, false).use { output ->
        output.write(bytes)
        output.flush()
        output.fd.sync()
      }
    }

    AsyncFunction("appendAndSync") { uriString: String, bytes: ByteArray ->
      val file = fileFromUri(uriString)
      file.parentFile?.mkdirs()
      FileOutputStream(file, true).use { output ->
        output.write(bytes)
        output.flush()
        output.fd.sync()
      }
    }

    AsyncFunction("replaceAndSync") { sourceUri: String, destinationUri: String ->
      val source = fileFromUri(sourceUri)
      val destination = fileFromUri(destinationUri)
      destination.parentFile?.mkdirs()
      Os.rename(source.absolutePath, destination.absolutePath)
      val directory = destination.parentFile ?: error("Atomic destination has no parent")
      val descriptor = Os.open(directory.absolutePath, OsConstants.O_RDONLY, 0)
      try {
        Os.fsync(descriptor)
      } finally {
        Os.close(descriptor)
      }
    }
  }

  private fun fileFromUri(value: String): File {
    val uri = Uri.parse(value)
    return when (uri.scheme) {
      "file" -> File(requireNotNull(uri.path) { "File URI has no path" })
      null, "" -> File(value)
      else -> error("Atomic files require an app-private file URI")
    }
  }
}
